File: coding/Leetcode/DS_A/task_253.cpp
// https://leetcode.com/problems/meeting-rooms-ii/description/
// Created by ruslan515 on 30.06.23.
//

#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    int minMeetingRooms(vector<vector<int>> &intervals) {

    }
};

int main() {
    vector<vector<int>> intervals;
    int answer;
    Solution solve;

    intervals = {{0,  30},
                 {5,  10},
                 {15, 20}};
    answer = 2;
    assert(answer == solve.minMeetingRooms(intervals));

    intervals = {{7, 10},
                 {2, 4}};
    answer = 1;
    assert(answer == solve.minMeetingRooms(intervals));

    intervals = {{13, 15},
                 {1,  13}};
    answer = 1;
    assert(answer == solve.minMeetingRooms(intervals));

    return 0;
}

File: coding/Leetcode/Tasks/task_253.py
# https://leetcode.com/problems/meeting-rooms-ii/?envType=weekly-question&envId=2024-03-15
from typing import List, Optional


class Solution:
    def minMeetingRooms(self, intervals: List[List[int]]) -> int:
        n = len(intervals)
        if n == 1:
            return 1
        answer = 0
        f = lambda x: x[0]
        intervals.sort(key=f)
        i = 0
        while i < n:
            second = intervals[i][1]
            j = i + 1
            while j < n and intervals[j][0] <= second:
                j += 1
                answer += 1
            if answer != 0:
                answer += 1
            i = j
        if answer == 0:
            answer = 1
        else:
            answer -= 1
        return answer


def main():
    solve = Solution()

    intervals = [[5, 8], [6, 8]]
    answer = 2
    assert answer == solve.minMeetingRooms(intervals)

    intervals = [[0, 30], [5, 10], [15, 20]]
    answer = 2
    assert answer == solve.minMeetingRooms(intervals)

    intervals = [[7, 10], [2, 4]]
    answer = 1
    assert answer == solve.minMeetingRooms(intervals)


if __name__ == "__main__":
    main()
